### DISCUSSION TITLE
Cannot cast audio to Google Chromecast

### DISCUSSION THREAD

**Discussion by roldengarm** [music-assistant/support#2209] [Q&A] [answered]
### What version of Music Assistant has the issue?

2.0.0b145

### What version of the Home Assistant Integration have you got installed?

n/a

### Have you tried everything in the Troubleshooting FAQ and reviewed the Open and Closed Issues and Discussions to resolve this yourself?

- [X] Yes

### The problem

I'm trying to cast Deezer to a Google Chromecast (one without Google TV). It finds the Chromecast, but when I select a song, nothing happens, counter stays at 0:00

I tried RadioBrowser, same problem.
Also tried a new CCWGTV, same problem. It shows "Music Assistant" on the TV, but no audio.

This is repeatedly in the logs:
```
2024-05-07 02:28:03.723 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/aiohttp/web_protocol.py", line 452, in _handle_request
    resp = await request_handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/aiohttp/web_app.py", line 543, in _handle
    resp = await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/controllers/streams.py", line 291, in serve_queue_item_stream
    async for chunk in get_ffmpeg_stream(
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/helpers/audio.py", line 803, in get_ffmpeg_stream
    async with FFMpeg(
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/helpers/process.py", line 80, in __aenter__
    await self.start()
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/helpers/audio.py", line 120, in start
    await super().start()
  File "/usr/local/lib/python3.12/site-packages/music_assistant/server/helpers/process.py", line 96, in start
    self.proc = await asyncio.create_subprocess_exec(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/subprocess.py", line 224, in create_subprocess_exec
    transport, protocol = await loop.subprocess_exec(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1744, in subprocess_exec
    transport = await self._make_subprocess_transport(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/unix_events.py", line 211, in _make_subprocess_transport
    transp = _UnixSubprocessTransport(self, protocol, args, shell,
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/base_subprocess.py", line 36, in __init__
    self._start(args=args, shell=shell, stdin=stdin, stdout=stdout,
  File "/usr/local/lib/python3.12/asyncio/unix_events.py", line 820, in _start
    self._proc = subprocess.Popen(
                 ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/subprocess.py", line 992, in __init__
    errread, errwrite) = self._get_handles(stdin, stdout, stderr)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/subprocess.py", line 1703, in _get_handles
    fcntl.fcntl(p2cwrite, fcntl.F_SETPIPE_SZ, self.pipesize)
PermissionError: [Errno 1] Operation not permitted
2024-05-07 02:28:03.767 WARNING (Thread-7) [music_assistant.providers.chromecast] Load media failed: 1 - error code: 104
```

### How to reproduce

Select a song.
Select a Chromecast to play
Nothing happens

### Music Providers

Deezer
RadioBrowser

### Player Providers

Chromecast

### Full log output

[log.txt](https://github.com/music-assistant/hass-music-assistant/files/15229030/log.txt)

### Additional information

Using self hosted on Docker.
Not integrated with HA.

### What version of Home Assistant Core are your running

2024.5.1

### What type of installation are you running?

Home Assistant Container

### On what type of hardware are you running?

Raspberry Pi

**Comment by OzGav** [comment]
Please follow the instructions and only attach - do not paste - the log so we can read this report.

**Comment by roldengarm** [comment]
> Please follow the instructions and only attach - do not paste - the log so we can't read this report.

Sorry about that @OzGav ! I've just done that.

**Comment by OzGav** [comment]
OK can you list what you have tried from the troubleshooting page and what didn't help.

**Comment by roldengarm** [comment]
> OK can you list what you have tried from the troubleshooting page and what didn't help.

I've tried restarting MA, cleared the cache (in MA), tried enable/disable Enable queue flow mode. It's showing as Flac, but for CC there doesn't seem to be a setting to use MP3.
Tried with both an old school Chromecast as well as a CC with GTV.

Tried two music providers, Deezer and RadioBrowser.

Checked the logs, which shows a 104 Chromecast error, based on my internet search it seems like a coding issue.

I've previously also tried a DLNA app on my tablet (BubbleUPNP) but that was causing similar issues.

I don't have a complicated network setup, all on the same subnet.

Just a side note, there is no feedback in the UI, i.e. it just does not play, it would be great if errors could be displayed in the UI.

Please let me know what else you need to know @OzGav

**Comment by OzGav** [comment]
Have you ever had MA working?

**Comment by roldengarm** [comment]
> Have you ever had MA working?

Not really. It worked briefly using DLNA, but was not reliable. Therefore trying to use Chromecast. @OzGav 
It keeps throwing that 104 error code in the logs.

**Comment by OzGav** [comment]
The log you supplied is showing beta 143? I see you are running in docker. Have you started the container in host networking mode and privileged mode?

https://github.com/music-assistant/hass-music-assistant/issues/2150

**Comment by roldengarm** [comment]
I'm using this image: `image: ghcr.io/music-assistant/server`

Setting privileged to true seems to make it work, sometimes. The MA UI still doesn't work all the time, randomly Music Assistant does show up blank on Chromecast, then I try a different song, and the MA UI doesn't update. The lack of error handling in the MA UI is frustrating. 

Sorry but in this state I can't hand it over to non tech people in the house. I've spent already a lot of time on this, think I'm going to give up.
Thanks for the prompt support @OzGav

**Comment by OzGav** [comment]
OK. We would like to get it working for you as there are 10000+ people using it these days and we are just about to go to the stable build.  You do have to run it in privileged mode. In that mode if there are different errors in the log then you can post that and we can have a look. Regardless I will leave this open for a while as Marcel may spot something straight away

**Comment by roldengarm** [comment]
My major concern with MA at this stage is the lack of error handling in the UI. 
I've shown my wife, and she tried, and immediately said it didn't work as it took ~10 seconds before it started playing. Then she tried to change the volume, which was applied with a ~5s second delay, so then it was suddenly blasting through the house.

Also, selecting another player is frustrating as it opens on the side and it's confusing how to select it.

I just tried, another error in the logs:
```
music-assistant-1  | 2024-05-07 05:15:56.294 WARNING (MainThread) [music_assistant.providers.deezer] Failed getting album: {'error': {'type': 'DataException', 'message': 'no data', 'code': 800}}
music-assistant-1  | 2024-05-07 05:15:56.299 WARNING (MainThread) [music_assistant.music.track] Unable to fetch album details deezer--NzGbVhyV://album/86389 - album://86389 not found on provider deezer--NzGbVhyV
music-assistant-1  | 2024-05-07 05:15:56.714 ERROR (MainThread) [music_assistant.webserver] Error handling message: players/queue/play_media: No playable items found
```
Had to try three times and then it suddenly worked.

**Comment by OzGav** [comment]
Yes there is something wrong with your setup as it works very fast for most people. What hardware are you running it on? How much RAM?

**Comment by roldengarm** [comment]
On a RPI4. Zero issues in other apps, like HA, it's very fast.

My issue is not really that it's slow, but that the UI doesn't show any progress indicators / errors / warnings.

**Comment by OzGav** [comment]
Progress is shown when you click on a track by the player bar showing the selected track, the button in the centre changes to a pause icon, the track counter goes to zero and starts counting up. This should happen within a second. When you change volume that is instantaneous. 

How much RAM does your Pi have?

**Comment by roldengarm** [comment]
4GB.

**Comment by OzGav** [comment]
It should work ok then unless you are running a lot of other addons. MA needs around 2GB RAM. 

Looking bck at another of your comments. Players are selected as you say by popping out the player menu. You then click on the one you want. The name of that player will then appear in the player bar so you know which one you are controlling. When you have multiple players playing simultaneously then you can see what they are all doing on the home screen and pause and resume them there. 

Why you got that error with Deezer I am not sure as I don't use that provider. Try sending some radio browser streams, that should work fine. Note there will be a slight delay which is outside of MA's control as the station has to start their stream after the request.

**Comment by roldengarm** [comment]
Thanks @OzGav  much appreciated. I will give it a try.

On a side note, unfortunately my plan to use a Chromecast as audio receiver failed. I bought an audio extractor (HDMI => SPDIF) that arrived today so I can plug it into my speakers (with bt / spdif / rca input). However, it still requires a TV to be connected. 
I also tried a tablet to serve as DLNA receiver via BubblePNP, but that requires a very beefy tablet by the looks of it, my Lenovo Tab4 couldn't handle it.
Also tried FullyKiosk as audio target; same problem. So I've run out of options to make my dumb speakers smart, unless you have a brilliant idea. 
It's a shame Google discontinued CC Audio.

**Comment by OzGav** [comment]
If you have a spare RPi (any version) you can run PiCorePlayer on it and that can connect via BT or the 3.5mm stereo jack. There are also ESP32 options. 

Yes if you ever find a CC audio in the $1 bin somewhere then grab it!

**Reply by roldengarm** [reply]
Hi @OzGav  hope you don't mind me asking, but I'm still struggling to figure out what to buy. I'd like a ready-made solution, as I don't have a 3D printer to create a custom box for an ESP device. I'd like to use the SPDIF or COAX in on my speakers.
I've been searching for a long time now, but I'm still unsure where to start.
A RPI only has 3.5mm jack / BT, so I'd need extra hardware for that as well, plus RPI's are hard to come by.
I did see the [docs](https://music-assistant.io/faq/stream-to/#music-assistant), but that didn't help me further.

[Squeezelite ](https://github.com/sle118/squeezelite-esp32)seems like a good idea, and it says you can connect a cable to connect to SPDIF, but the ESP32 board doesn't have SPDIF out so that's again confusing :(

Also not keen to spend hundreds of dollars on this, based on the fact it should be relatively cheap with an ESP32 device.

**Reply by OzGav** [reply, marked as answer]
If you want to come to the lounge on Discord there has just been a bit of discussion about that and one of the guys has been asked to write uo something that answers your question. That extended response won't be done until the end of the month but you should be able to find some useful info there now.

P.S. You have the HDMI output from the RPi as well

**Reply by roldengarm** [reply]
Thanks @OzGav ! 
Glad that documentation is being worked on.
I've discovered that it works very well via the Fully integration. Previously it didn't work well, perhaps because I had not enabled privileged.

**Comment by erkr** [comment]
I use CC videos with an active powered hdmi audio splitter by lack of CC audios. The CC can't drive passive splitters. They don't need a TV set. I think this is only for passive adapters as the TV provides the power

**Comment by erkr** [comment]
I just had a very similar issue. Music was stopped, casting the MA random 100 playlist. 
The queue was far from done. Only Spotify provider and Chromecast players. I checked: no errors logged.

Then I noticed probably the same issue: pressing play did not resume the playlist. It just jumped to the next track and remained in stopped state. No errors logged again.

I 'solved' it by clearing the queue and start a new one

Also running RPi 4, normally no issues like this.

**Comment by roldengarm** [comment]
> I use CC videos with an active powered hdmi audio splitter by lack of CC audios. The CC can't drive passive splitters. They don't need a TV set. I think this is only for passive adapters as the TV provides the power

I have got an active splitter from Temu, but it only works when there's a TV connected to the output @erkr

**Comment by erkr** [comment]
@roldengarm 
I bought these previous on ali:
https://a.aliexpress.com/_EGYH223
The ones I got work without TV

**Comment by OzGav** [comment]
I;m just converting this to a discussion as we seem to have veered away from this being a MA issue. I am assuming you are running the docker container correctly now and that is all fine. For all your testing I would suggest you use the Chromecast. If that is all working fine then we can look at providing more advice about a solution for your speakers. If anything isn't working then let us know and we will tackle one problem at a time.

**Reply by Zerwin** [reply]
I still have the same (or a similar) problem.
Logs show `WARNING (Thread-6) [music_assistant.providers.chromecast] Load media failed: 1 - error code: 104`, and on the TV I can see that Music Assistant tries to play the music but the progress bar is just infinitely looping. I also tried with different providers, both Youtube Music and Local files in different file formats, but all the same result.

**Reply by OzGav** [reply]
What about other players. Do they work ok?

**Reply by roldengarm** [reply]
@Zerwin are you running on Docker? If so, please check you have enabled privileged as that was the issue in my case.

**Reply by Zerwin** [reply]
I am running it in Home Assistant as an Addon. I tried using Snapweb as a client, and while it showed up in MA as a Player, it just was on idle and I couldn't select it to actually play any music. MA refused to let me select it and because I had no speaker selected I couldn't Queue any music.

**Reply by OzGav** [reply]
Sounds like a network problem. Have you been through the troubleshooting page in the docs?

**Reply by Zerwin** [reply]
(Unrelated, but the issue link at https://music-assistant.io/faq/troubleshooting/ under the "My media player is not available or not playing" does not work, it points to https://github.com/orgs/music-assistant/issues )

So I got Snapweb to work now, not sure what exactly it was. 
At first I tried with a local file and that worked, although the quality was a bit low.
Then I tried with Youtube Music as Provider, however, it seems like that crashed the Snapcast server ?

```
2024-05-16 02:33:01.506 DEBUG (MainThread) [music_assistant.players] Handling command cmd_stop for player Snapweb client
2024-05-16 02:33:01.525 DEBUG (MainThread) [music_assistant.streams.media_stream] stream aborted (with code None) for ytmusic--5TSofULq://track/oc2k5F5uGn0 - seconds streamed: 6.0
2024-05-16 02:33:01.556 DEBUG (MainThread) [snapcast.control.group] set stream to default on Snapweb client
2024-05-16 02:33:01.591 DEBUG (MainThread) [music_assistant.providers.snapcast.ffmpeg] Process ffmpeg with PID 150 stopped with returncode 255
2024-05-16 02:33:01.592 DEBUG (MainThread) [music_assistant.providers.snapcast] Finished streaming to tcp://127.0.0.1:4953
2024-05-16 02:33:01.613 DEBUG (MainThread) [music_assistant.providers.snapcast.snapserver] 2024-05-16 02-33-01.593 [Info] (Server) Stream.RemoveStream("MusicAssistant--4953")
2024-05-16 02:33:01.614 DEBUG (MainThread) [music_assistant.providers.snapcast.snapserver] 2024-05-16 02-33-01.594 [Error] (AsioStream) Error during async wait: Operation canceled
2024-05-16 02:33:02.342 DEBUG (MainThread) [snapcast.control.server] Server disconnected: [Errno 104] Connection reset by peer
2024-05-16 02:33:02.343 DEBUG (MainThread) [snapcast.control.server] try reconnect
2024-05-16 02:33:02.352 DEBUG (MainThread) [music_assistant.helpers.process.snapserver] Process snapserver with PID 24 stopped with returncode -11
2024-05-16 02:33:07.354 DEBUG (MainThread) [snapcast.control.server] try reconnect
```

I restarted both MA and Snapweb and after that it started working again. And now I can play music, both local and Youtube Music, via Snapcast.

However, Chromecast'ing to my Chromecast still didn't work, same behavior as before.

**Reply by OzGav** [reply]
You aren’t providing any info about why you don’t think this is a network problem

**Reply by Zerwin** [reply]
This was more to answer your previous question if other players work. Also I can cast from any other device to the Chromecast except MA. It works from my phone, my pc, and from Home Assistant I can even control the Chromecast, just MA doesn't work. If it was a network problem wouldn't that be universal ?

**Reply by marcelveldt** [reply]
For docker setups, in 99% of the cases if cast playback is not working, the base URL is incorrect in the settings --> core --> webserver.

**Reply by Zerwin** [reply]
Did you mean in settings --> core --> streamserver ? Under webserver I could only find the option to "Expose the Webserver". In streamserver the correct ip was set, I also tried with the internal hostname and FQDN, but with both I saw the same behaviour as before. I also tried binding to the actual IP of Home Assistant instead of 0.0.0.0 but still no luck. At last I exposed the webserver to directly access it via IP and then cast but that also didn't work. The Chromecast always looks like it's about to start playing but then just has an infinite spinner shown with the song title.

**Reply by Zerwin** [reply]
Okay, I figured it out, it was a firewall problem on my router. My Home Assistant was allowed to talk to my Chromecast, but the Chromecast wasn't allowed to talk to Home Assistant. Once I allowed that, it started working. Hope this helps someone else if they have the same problem.

Btw, it seems like the Sample Rates for Speakers have a typo, 48kHz 16bits is shown two times, I'm guessing one of these should be 24bits
![image](https://github.com/music-assistant/hass-music-assistant/assets/46265592/62022f9e-ffa0-4471-bf2e-bff728564c1a)

Also, should MA when it's casting to a Chromecast device show the cover image of the song playing or not ? Currently it doesn't for me but I'm not sure if that is a bug or intended.

**Reply by OzGav** [reply]
Thanks at this stage of MA it usually is a network problem.

someone else mentioned the duplicated sample rate so we will look at that 

it might depend on the chromecast device. Try with flow mode on and off.

**Reply by mikeb9x** [reply]
Is there a way to manually enter the IP of a Chromecast player so MA won't change it? MA is frequently losing connection with one of my speakers (it still works in HA and Google Home app) and seems like the issue might be with the IP - when the speaker is unavailable the IP in MA is changed to something like [::ffff:c0a8:22]:8009, where it should be normal IPv4 address (I have IPv6 turned off in my router).

**Reply by OzGav** [reply]
No that is not possible. MA will not be changing the IP. I will check what the recommendation is for IPv6

**Reply by madbrain76** [reply]
> Is there a way to manually enter the IP of a Chromecast player so MA won't change it? MA is frequently losing connection with one of my speakers (it still works in HA and Google Home app) and seems like the issue might be with the IP - when the speaker is unavailable the IP in MA is changed to something like [::ffff:c0a8:22]:8009, where it should be normal IPv4 address (I have IPv6 turned off in my router).

Have you tried disabling IPv6 in HA  under Settings -> System -> Network ? Not sure if this will affect MA, but it might. I left this setting as "Automatic", but perhaps "Disabled" will help in your case.

I also have IPv6 disabled in my router, because my ISP doesn't support IPv6 yet.

All my Chromecast audio devices and WiiM Pro Plus with built-in CCAs have DHCP reservations, so the IPv4 addresses don't change. I just noticed that the WiiM app also has an "Enable IPv6" device option under Network status / Configure IP. I also left his option at the default of enabled for all 12 WiiM devices.

For the 2 actual Google-branded CCAs that I still have, there is no IPv6 setting. However, the Google Home app shows the IPv4 address only in settings / device information. Same for the video devices - 2 Chromecast with Google TV (4K version), and the Chromecast Ultra. No trace of IPv6 anywhere.

It all seems to work fine with IPv4 only, so far.
